### PR TITLE
Fix Issue 18280 - std.algorithm.comparison.cmp for non-strings should call opCmp only once per item pair

### PR DESCRIPTION
**Pull request by n8sh**
Right now when comparing non-string ranges std.algorithm.comparison.cmp calls the comparison predicate twice for every pair of elements to determine their order. When the elements have overloaded `opCmp` and the predicate is "a < b" we only need to call `opCmp` once for each pair.

### CODE REVIEW

File: std/algorithm/comparison.d [hunk 1]
```diff
@@ -610,13 +610,26 @@ if (isInputRange!R1 && isInputRange!R2)
 {
     static if (!(isSomeString!R1 && isSomeString!R2))
     {
+        static if (is(typeof(pred) : string))
+            enum isLessThan = pred == "a < b";
```
Comment (wilzbach) [inline std/algorithm/comparison.d]: FYI: @andralex isn't a huge fan of string comparisons, see https://github.com/dlang/phobos/pull/4265#issuecomment-266234691

File: std/algorithm/comparison.d [hunk 2]
```diff
@@ -610,13 +610,26 @@ if (isInputRange!R1 && isInputRange!R2)
...
         for (;; r1.popFront(), r2.popFront())
         {
             if (r1.empty) return -cast(int)!r2.empty;
             if (r2.empty) return !r1.empty;
-            auto a = r1.front, b = r2.front;
-            if (binaryFun!pred(a, b)) return -1;
-            if (binaryFun!pred(b, a)) return 1;
+            static if (isLessThan && is(typeof(r1.front.opCmp(r2.front)) : int))
+            {
+                immutable int c = r1.front.opCmp(r2.front);
```
Comment (wilzbach) [inline std/algorithm/comparison.d]: Why not `r1.front < r2.front`?

File: std/algorithm/comparison.d (hunk 1 above)
Comment (wilzbach) [reply std/algorithm/comparison.d]: As I did a bit of digging the string specialization of `cmp` was added in 2011: https://github.com/dlang/phobos/commit/a0ecf2a10e319e387e269fd6310ad2acd244cc92

Comment (andralex) [reply std/algorithm/comparison.d]: Good news, https://github.com/dlang/dmd/pull/7484/ is on. With it we should be able to switch to lambdas soon. cc @RazvanN7

File: std/algorithm/comparison.d [hunk 3]
```diff
@@ -689,6 +698,54 @@ if (isInputRange!R1 && isInputRange!R2)
...
 }
 
+/// ditto
+int cmp(alias pred, R1, R2)(R1 r1, R2 r2)
+if (isInputRange!R1 && isInputRange!R2)
+{
+    static if (is(typeof(pred) : string)
+        && (__traits(compiles, { static assert("a < b" == pred); })))
+    {
+        // In case someone was explicitly using the default predicate.
```
Comment (andralex) [inline std/algorithm/comparison.d]: please remove, no need to accommodate this case

Comment (n8sh) [reply std/algorithm/comparison.d]: Removed

**Comment by n8sh** [comment]
>I committed a few changes.

Sorry, I clobbered your changes when I force pushed to add a unittest.

**Comment by andralex** [comment]
OK, I submitted them again.

I think since we got here we should fix the `float` return. So cmp should return auto and let opCmp drive.

**Comment by n8sh** [comment]
For float `opCmp` what do we do with a NaN result? "Not positive or negative, so we keep going" or "not zero, so we return it"?

**Comment by andralex** [comment]
@n8sh Stop the comparison and forward the NaN result. So the core of the loop would be:

```d
auto c = r1.front.opCmp(r2.front);
if (c != 0) return c;
```

File: std/algorithm/comparison.d [hunk 4]
```diff
@@ -614,40 +623,41 @@ if (isInputRange!R1 && isInputRange!R2)
         {
             if (r1.empty) return -cast(int)!r2.empty;
             if (r2.empty) return !r1.empty;
```
Comment (andralex) [inline std/algorithm/comparison.d]: For these two returns you need a bit of massaging. I'd say something like this:

```d
static if (is(typeof(r1.front.opCmp(r2.front))) R)
    alias Result = R;
else
    alias Result = int;
if (r1.empty) return -R(!r2.empty);
if (r2.empty) return R(!r1.empty);
...
```

File: std/algorithm/comparison.d [hunk 5]
```diff
@@ -724,6 +776,60 @@ if (isInputRange!R1 && isInputRange!R2)
...
+@safe unittest
+{
+    int result;
+
+    result = cmp!"a > b"("abc", "abc");
+    assert(result == 0);
+    result = cmp!"a > b"("", "");
+    assert(result == 0);
+    result = cmp!"a > b"("abc", "abcd");
+    assert(result > 0);
```
Comment (n8sh) [inline std/algorithm/comparison.d]: This behavior is a bug. r1 is a prefix of r2, so the result should be negative even if we are comparing in reverse alphabetical order, but `cmp` is using `pred` to compared the lengths! I have no sense of whether anyone might have been relying on this for reverse order comparison. It is a violation of the documentation and it only works for strings. Should it be fixed?

File: std/algorithm/comparison.d (hunk 4 above)
Comment (n8sh) [reply std/algorithm/comparison.d]: Fixed

File: std/algorithm/comparison.d (hunk 5 above)
Comment (andralex) [reply std/algorithm/comparison.d]: That's terrible. I'd say open a bug with this repro then mark it as fixed by this PR. That way we'll have it in the changelog.

**Comment by andralex** [comment]
@n8sh you'll need to add a few more unittests because now you are generalizing cmp. Sorry this fell on you and thanks for the quality code!

**Comment by n8sh** [comment]
Found another bad bug. This code:
```if (c1 != c2) return threeWayInt(cast(int) c1, cast(int) c2);```
meant that no custom predicate that treats different values the same (for example, ignoring case) will work correctly. They'll incorrectly return 0 prematurely.

Example https://run.dlang.io/is/feq4KK
```d
void main()
{
    import std.algorithm.comparison : cmp;
    static bool lessThanCaseInsensitive(size_t a, size_t b)
    {
        import std.ascii : toUpper;
        return toUpper(cast(dchar) a) < toUpper(cast(dchar) b);
    }
    static assert(cmp!lessThanCaseInsensitive("apple2", "APPLE1") != 0,
                 "These are clearly not the same!");
}
```

**Comment by andralex** [comment]
@n8sh it's a shame these haven't been exposed already. Thanks for your work! Please file that bug, too (or massage it with the other) and then mark them as fixed with this PR.

**Comment by andralex** [comment]
@n8sh I made a push to your code that removes `cast` in promotions.

**Comment by andralex** [comment]
Approved, please squash and then we'll merge. Thanks!!

**Comment by andralex** [comment]
Aaand made one more minor optimization - `empty` was sometimes checked twice at the end of the ranges.

**Comment by ZombineDev** [comment]
@n8sh perhaps the way forward is to make it an option but also keep the existing behavior by querying the predicate type via static if.

**Comment by andralex** [comment]
There's always another name such as `cmp3`... but I'd say distinguishing a predicate that returns `bool` from one that returns anything else is a fine option too.

**Comment by aG0aep6G** [comment]
The bot picked up 4 issues here, but it only closed one. Bug?
https://github.com/dlang-bots/dlang-bot/issues/175

**Comment by n8sh** [comment]
After @aG0aep6G's message I closed the issues manually, but I don't know if that might result in them not getting added to the next release changelog.

**Comment by wilzbach** [comment]
> After @aG0aep6G's message I closed the issues manually, but I don't know if that might result in them not getting added to the next release changelog.

A bit of explanation:

The bugs are closed by the Bugzilla integration - not the bot.
The changelog is generated by https://github.com/dlang/tools/blob/master/changed.d - it parses the git log between two revisions with the same regex as the bot and the bugzilla integration does. However, it then queries Bugzilla and filters for only referenced bugs which are closed.
So the bug overview by the bot is merely a help for reviewers as (a) it provides a handy link to the references issues and (b) allows them to verify that the issues have been referenced correctly.
See also: https://github.com/dlang-bots/dlang-bot#automated-references

Now to the bug - it looks like the Bugzilla integration is a bit dumber and only picks up one bug regex per commit. There's much I can do about (except for writing our own Bugzilla integration).
In any case, a "workaround" for the future if you have to use one commit is to include multiple issues in one regex, e.g. "Fixes issues 17494, 17505, 17506".

**Pull request merged**